Matt Parker's Math Puzzle #1
======
Since Matt Parker love some brute force python, I propose this solution his Math Puzzle #1.

https://www.youtube.com/watch?v=T29dydI97zY

Import
------
To ease our task, let's use some librairie
### itertools
itertools let us use the permutation fonction to generate all the posible way our investor can sit around the table.
### pandas
pandas let us manipulate data more easily in a table format

In [1]:
import itertools
import pandas as pd

Is Sit Properly
-------
We define a function that let us check if our investor are sit on the right seat.

### How does it work
For each chair we look if it's the right person and store the result in a list.
At the end, we combine all those result to get back a table similar to our seat list but with boolean value

We also allow to pass an Offset value to simulate the table rotation

In [2]:
def IsSitProperly(pdData,ListOfPerson,Offset=0):
    lResult = []
    for personIndex, chair in enumerate(ListOfPerson):
        lResult.append(pdData[chair]==ListOfPerson[(personIndex+Offset)%len(ListOfPerson)])
    return pd.DataFrame(lResult)

Defining constant
------
We define our constant:
* **NbOfPerson** : Number of investor
* **NbToWin** : Number of investor needed to get our funding
* **NbToStart** : This define the number of investor sitted on the right chair just before spinning the table

In [11]:
NbOfPerson = 7
NbToWin = 2
NbToStart = 1

Main data generation
------
From there, we generate the list of investor and all the permutation

With 7 investor, there's 5040 permutation

In [12]:
ListOfPersons = list(range(1,NbOfPerson+1))
AllPermutations = itertools.permutations(ListOfPersons, NbOfPerson)
pdPermutations = pd.DataFrame(list(AllPermutations))
pdPermutations.columns = ListOfPersons
print(len(pdPermutations), "permutation")
pdPermutations

5040 permutation


1  2  3  4  5  6  7
0     1  2  3  4  5  6  7
1     1  2  3  4  5  7  6
2     1  2  3  4  6  5  7
3     1  2  3  4  6  7  5
4     1  2  3  4  7  5  6
...  .. .. .. .. .. .. ..
5035  7  6  5  4  1  3  2
5036  7  6  5  4  2  1  3
5037  7  6  5  4  2  3  1
5038  7  6  5  4  3  1  2
5039  7  6  5  4  3  2  1

[5040 rows x 7 columns]

Starting Situation
------
With all our permutation, we get a list of permutation that respect the starting condition

In [13]:
StartingSituation = pdPermutations.loc[IsSitProperly(pdPermutations,ListOfPersons).sum()==NbToStart]
print("Starting with ", len(StartingSituation), "permutation")
StartingSituation

Starting with  1855 permutation


1  2  3  4  5  6  7
127   1  3  2  5  4  7  6
129   1  3  2  5  6  7  4
130   1  3  2  5  7  4  6
133   1  3  2  6  4  7  5
136   1  3  2  6  7  4  5
...  .. .. .. .. .. .. ..
5035  7  6  5  4  1  3  2
5036  7  6  5  4  2  1  3
5037  7  6  5  4  2  3  1
5038  7  6  5  4  3  1  2
5039  7  6  5  4  3  2  1

[1855 rows x 7 columns]

Solution
======
From there, for each table rotation, we check if a permutation is a winner.

At the end, we find all the permutation where there's no spin that let us win.

For the problem expose by Matt, there's **133** starting position where there's no win possible

In [14]:
RotationResult = []
for TableRotation in range(1,NbOfPerson):
    RotationResult.append(IsSitProperly(StartingSituation,ListOfPersons,TableRotation).sum()>=NbToWin)
pdResult = StartingSituation.loc[pd.DataFrame(RotationResult).sum()==0]
pdResult

1  2  3  4  5  6  7
186   1  3  5  7  2  4  6
197   1  3  6  2  7  5  4
236   1  3  7  6  4  2  5
262   1  4  2  7  6  3  5
319   1  4  6  3  2  7  5
...  .. .. .. .. .. .. ..
4869  7  5  3  6  2  4  1
4875  7  5  4  1  3  6  2
4987  7  6  3  5  1  4  2
5002  7  6  4  2  5  1  3
5039  7  6  5  4  3  2  1

[133 rows x 7 columns]